### ISSUE TITLE
Not sure what's going on with this error.

### ISSUE THREAD

**Issue by AndyT86** [Alex079/TinySuite#4]
Try to use TinyTimer.h I #include nothing else, so some #ifndef is missing? But that can't be because nothing else is #included. And on another note, no actual example is around here. You need to do something there.

wiring.c.o (symbol from plugin): In function `__vector_5':
(.text+0x0): multiple definition of `__vector_5'
libraries\TinySuite\TinyTimer0Overflow.cpp.o (symbol from plugin):(.text+0x0): first defined here
collect2.exe: error: ld returned 1 exit status
exit status 1
Error compiling for board ATtiny25/45/85.

**Comment by Alex079** [comment]
Hello!
By looking at the error message you provided, I see that wiring.c is included in the build scope.
I found a similar problem described here https://arduino.stackexchange.com/questions/70126/arduino-ide-with-attiny85-error-multiple-definition-of-vector-5

I checked the example for timer overflow https://github.com/Alex079/TinySuite/blob/master/examples/timer-overflow/timer-overflow.ino
Indeed, it can fail with this error if Arduino core is using Timer0/Timer1 internally (I assume you are using Arduino with ATTinyCore).

Please try disabling "millis()/micros()" feature if you are using a compatible version of ATTinyCore and a corresponding board definition.
Alternatively, you can try creating `int main() {...}` function directly instead of using `void setup() {...}` + `void loop() {...}` provided by Arduino.

I hope this helps.
Regards,
Alex.

**Comment by AndyT86** [comment]
Got that fixed, but I must not have used it correctly, code loaded up and RELAY pin is high as expected but LED pin is off which isn't right, timer is not running.

#include <TinyTimer.h>


const int RELAY_OUT_PIN = 2;
const int PIR_IN_PIN = 0;
const int LED_PIN = 3;

void timer_setup() {
}

void timer_teardown() {
}

void timer_click() {
  digitalWrite(LED_PIN, HIGH);   // turn the LED on (HIGH is the voltage level)
}
void func() {
  digitalWrite(RELAY_OUT_PIN, HIGH);
}

// the setup function runs once when you press reset or power the board
int main() {
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_PIN, OUTPUT);
  pinMode(RELAY_OUT_PIN, OUTPUT);
  pinMode(PIR_IN_PIN, INPUT);
  TinyTimer tt = TinyTimer(timer_setup, timer_teardown);
  tt.on(2000, timer_click);
  tt.onTimer = timer_click;
  while (1) {
    digitalWrite(RELAY_OUT_PIN, HIGH);
    //delay(2000);
  }
}

**Comment by AndyT86** [comment]
Uh ATtinyCore? All that stuff went over my head, I only got my ATtiny85 in the mail yesterday; it is a minor miracle to have done the ISP right in 1 go.

**Comment by AndyT86** [comment]
Still not working.

#include <TinyTimer.h>

#include <avr/interrupt.h>

const int RELAY_OUT_PIN = 2;
const int PIR_IN_PIN = 0;
const int LED_PIN = 3;

void timer_setup() {
}

void timer_teardown() {
}

void timer_click() {
  digitalWrite(LED_PIN, HIGH);   // turn the LED on (HIGH is the voltage level)
}
void func() {
  digitalWrite(RELAY_OUT_PIN, HIGH);
}

// the setup function runs once when you press reset or power the board
int main() {
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_PIN, OUTPUT);
  pinMode(RELAY_OUT_PIN, OUTPUT);
  pinMode(PIR_IN_PIN, INPUT);
  sei();
  TinyTimer tt = TinyTimer(timer_setup, timer_teardown);
  tt.on(2000, timer_click);
  tt.onTimer = timer_click;
  while (1) {
    digitalWrite(RELAY_OUT_PIN, HIGH);
    //delay(2000);
  }
}

**Comment by Alex079** [comment]
OK. This looks a bit confusing.
When you `#include <TinyTimer.h>` you will have to implement a timer yourself (provide meaningful setup/teardown functions and make callback function invoked). The timer in the code above does not work, because there is no timer implementation provided.

I made four timer implementations for regular usage of the two hardware timers of ATTiny85:
|#include|ISR which will be included|
|-|-|
|TinyTimer0Compare|TIMER0_COMPA_vect|
|TinyTimer1Compare|TIMER1_COMPA_vect|
|TinyTimer0Overflow|TIMER0_OVF_vect|
|TinyTimer1Overflow|TIMER1_OVF_vect|

You can include any of these headers, as long as there is no conflict with other included libraries which may use the same hardware timers.
`onTimer` does not need to be set explicitly (although it can be). I had to make it public for ISRs to be able to invoke it.

I guess I should put this in a bit more details in README.

The ATTinyCore is this https://github.com/SpenceKonde/ATTinyCore

**Comment by AndyT86** [comment]
I am lost, AVR-C calls for a complete reading of datasheet as well as bitwise mastery in C, I know you were doing something to Pin B0 and B1 in the Compare example but that's the extend of my understanding.

**Comment by Alex079** [comment]
Maybe this example will work better. Please note that the timer delay is in CPU clock tick count and for now is limited to 65535 ticks max. Default F_CPU for ATTiny85 is 8 MHz as far as I remember, this gives about 8 ms delay max.
```
#include <TinyTimer0Overflow.h>

const int RELAY_OUT_PIN = 2;
const int PIR_IN_PIN = 0;
const int LED_PIN = 3;

const uint16_t TIMER_DELAY_IN_CPU_TICKS = 65535;

void timer_click() {
  digitalWrite(LED_PIN, HIGH); // turn the LED on (HIGH is the voltage level)
}

int main() {
  pinMode(LED_PIN, OUTPUT);
  pinMode(RELAY_OUT_PIN, OUTPUT);
  pinMode(PIR_IN_PIN, INPUT);
  digitalWrite(RELAY_OUT_PIN, HIGH);
  sei(); // set "enable interrupts"
  Timer0Overflow.on(TIMER_DELAY_IN_CPU_TICKS, timer_click);
  while (1) {
    // do nothing
  }
  return 0;
}
```

**Comment by Alex079** [comment]
What happens here `PORTB ^= (1 << PB0)` is the toggle of pin 0 state. Maybe it is too low-level for an Arduino example, indeed. I will try to come up with more readable examples.

**Comment by AndyT86** [comment]
Oh that TinyCore thing I am sure I am not using it, it offers a lot but that a lot happens to be stuff I don't use, I bought a breakout just in case but it is at the moment using an Uno as ISP; the F_CPU of Tiny85 is 1, not 8 since bunch of YT says 1 MHz. And is "friend" keyword not usable in C++ here for that onTimer thing? Another part in tangent is interrupt, how do I do it? I know INT0 is on PB2 so far. You did not have it as library so can you point me to the right tutorial? Remember I got ATtiny85 only few days ago easy on the skill dump.

**Comment by AndyT86** [comment]
Here's something else just pop up, even stranger than me not knowing ATtiny85 enough, my no-timer-no-int code uploads to the tiny MCU just fine in parallel:

Sketch uses 1354 bytes (16%) of program storage space. Maximum is 8192 bytes.
Global variables use 149 bytes (29%) of dynamic memory, leaving 363 bytes for local variables. Maximum is 512 bytes.
avrdude: ERROR: address 0x820003 out of range at line 87 of C:\Users\actsa\AppData\Local\Temp\arduino_build_667981/timer-overflow.ino.hex
avrdude: read from file 'C:\Users\actsa\AppData\Local\Temp\arduino_build_667981/timer-overflow.ino.hex' failed
the selected serial port avrdude: read from file 'C:\Users\actsa\AppData\Local\Temp\arduino_build_667981/timer-overflow.ino.hex' failed
 does not exist or your board is not connected

**Comment by AndyT86** [comment]
Oh that's your example that has that error, while my basic code that you modified to use comparison timer works, fine, doing the switching very fast so a lot to be done, I have no idea how to change timing, probably the prescaler, and I might change to overflow since I need a long timer for timeout.

**Comment by Alex079** [comment]
Thank you for your help!

The `ERROR: address 0x820003 out of range` looks like Arduino IDE is trying to upload fuses into the same memory area as main program. I will remove the FUSES from the example.

Concerning the timer delay being too short, I will change the timer class to accept the maximum possible delay.

I made this branch to address some of the problems https://github.com/Alex079/TinySuite/tree/improve-timers
If you have time to test you can update the lib in your Arduino libraries folder.
If it works well for you I will update docs and make a new release.

Concerning the INT0, indeed, there is no wrapper class in the lib. The most detailed tutorial is, of course, the datasheet, but you can request a feature and I will implement it. Or you can take a look at `TinyPinChangeB` or search the net for `ISR(INT0_vect)`.

**Comment by AndyT86** [comment]
Sure, I bent my ATtiny85's pins, turns out I pushed into breadboard wrong; now I'm afraid to mess with it, will have to wait for my 2nd ATtiny to come in; or even for the ATmega. Anything you can see to make it not to mess around that 250+ page datasheet I would vote for it. Having to do that is just as close as doing assembly. Few can do it, but most cannot, and all would rather not to.

**Comment by Alex079** [comment]
I added a draft of INT0 wrapper here https://github.com/Alex079/TinySuite/tree/improvements
I did not test it yet.

**Comment by AndyT86** [comment]
You hit the nail right on, I want to do timer and interrupt precisely wanting the main thread into sleep. Me finding my chip and do this.

**Comment by Alex079** [comment]
I released the version 2.1.0, it should be deployed soon.
